## Single-nucleotide-polymorphism prediction
This script will map metagenomic reads to bacterial reference genomes and call SNPs in these genomes. You can either target one or more specific species, or provide this script a species abundance file  

The pipeline can be broken down into three main steps:  
  1) build a database of genome sequences for abundant bacterial species  
  2) map metagenomic reads to the database  
  3) use mapped reads to call SNPs  


## Usage
```
Usage: run_phylo_cnv.py snps outdir [options]

positional arguments:
  outdir                Path to directory to store results. Name should correspond to sample identifier.

optional arguments:
  -h, --help            show this help message and exit
  --remove_temp         remove intermediate files generated by PhyloCNV. 
                        intermediate files can be useful to quickly rerun parts of pipeline.

Pipeline options (choose one or more; default=all):
  --build_db            Build bowtie2 database of pangenomes
  --align               Align reads to pangenome database
  --call_snps           Run samtools mpileup and call SNPs

Database options (if using --build_db):
  --sp_cov GC_COV       Include species with >X coverage (3.0)
  --sp_topn GC_TOPN     Include top N most abundant species
  --sp_id GC_ID         One or more species identifiers to include in database. Separate ids with a comma

Read alignment options (if using --align):
  -1 M1                 FASTA/FASTQ file containing 1st mate if paired or unpaired reads
  -2 M2                 FASTA/FASTQ file containing 2nd mate if paired
  -s {very-fast,fast,sensitive,very-sensitive}
                        Bowtie2 alignment speed/sensitivity (very-sensitive)
  -n MAX_READS          # reads to use from input file(s) (use all)
  -t THREADS            Number of threads to use

SNP calling options (if using --call_snps):
  --mapid MAPID         Discard reads with alignment identity < MAPID (94.0)
  --mapq MAPQ           Discard reads with mapping quality < MAPQ (20)
  --baseq BASEQ         Discard bases with quality < BASEQ (30)
  --readq READQ         Discard reads with mean quality < READQ (20)
  --trim INT            Trim N base-pairs from read-tails (0)
  --baq                 Enable BAQ (per-base alignment quality)
  --redo_baq            Recalculate BAQ on the fly
  --adjust_mq           Adjust MAPQ
```

## Example

1) run entire pipeline using defaults:  
`run_phylo_cnv.py snps /path/to/outdir -1 /path/to/reads_1.fq.gz -2 /path/to/reads_2.fq.gz`
			
2) run entire pipeline for a specific species:  
`run_phylo_cnv.py snps /path/to/outdir --sp_id 57955 -1 /path/to/reads_1.fq.gz -2 /path/to/reads_2.fq.gz`

3) just align reads, use faster alignment, only use the first 10M reads, use 4 CPUs:  
`run_phylo_cnv.py snps /path/to/outdir --align -1 /path/to/reads_1.fq.gz -s very-fast -n 10000000 -t 4`  

4) just call SNPs, keep reads with >=95% alignment identity and keep bases with quality-scores >=35, and map quality scores >=20:  
`run_phylo_cnv.py snps /path/to/outdir --call_snps --mapid 95 --mapq 20 --baseq 35`

## Output

The output of this script is a directory with the following files:  

* **genomes.bam** (*intermediate output*): alignments of metagenomic reads versus representative genome(s)  
* **db/** (*intermediate output*): contains fasta and bowtie2 database containing genomes  
* **genomes.vcf** (*intermediate output*): vcf file for all genomes  
* **snps/** (*final output*): directory containing predicted SNPs for each selected species, 
* **snps_summary_stats.txt** (*final output*): alignment summary statistics for each species
* Intermediate outputs can be automatically removed by using the --remove_temp flag

Example polymorphism table for one sample (ex: snps/57955.snps.gz):

| ref_id      | ...      | ref_freq  |
| :----------: |:-------------:| :------------------: |
| accn|...         | ...        | 0.20              |
| ...           | ...           |   ...               |
| accn|...         | ...          |   0.35              |


* **ref_id**: scaffold id
* **ref_pos**: position on scaffold
* ref_allele: reference allele
* alt_allele: alternate allele
* cons_allele: consensus allele
* count_alleles: number of alleles observed in metagenome
* count_ref: count reference alleles observed
* count_alt: count alternate alleles observed
* depth: count total reads at ref_pos
* ref_freq: frequency (0.0 to 1.0) of reference allele

## Memory usage  
* Memory usage will depend on the number of species you search and the number of reference genomes sequenced per species.
* In practice, peak memory usage will not exceed 3.5 Gb for most samples

## Speed
* Speed will depend on the number of species you search and the number of sequenced reference genomes per species.
* For a single species with 1 reference genome, expect ~16,000 reads/second
* Use `-n` and `-t` to increase throughput

## Next step
[Merge results across samples] (https://github.com/snayfach/PhyloCNV/blob/master/docs/merge_snvs.md)